src/GraphQL/Internal/Validation.hs

```diff
@@ -625,6 +626,23 @@ data ValidationError
   | InvalidDefaultValue AST.Value
   deriving (Eq, Show)
 
+instance GraphQLError ValidationError where
+  formatError (DuplicateOperation name) = "More than one operation named '" <> show name <> "'"
+  formatError (MixedAnonymousOperations n names)
+    | n > 1 && null names = "Multiple anonymous operations defined. Found " <> show n
+    | otherwise = "Document contains both anonymous operations (" <> show n <> ") and named operations (" <> show names <> ")"
+  formatError (DuplicateArgument name) = "More than one argument named '" <> show name <> "'"
+  formatError (DuplicateFragmentDefinition name) = "More than one fragment named '" <> show name <> "'"
```

Comment:
One of the lessons learned from the psql client libraries was that it's useful to have a unique name somewhere in the `Error`, because if you don't people start matching strings to detect the error type, making the string representation your API. I'm not sure how we can fit this into `Error` but I think I'd like to keep that in mind.